Client/Frontend/Library/BookmarkDetailPanel.swift

```diff
@@ -0,0 +1,338 @@
+
+    // Editable field(s) that all BookmarkNodes have.
+    var parentBookmarkFolder: BookmarkFolder
+
+    // Editable field(s) that only BookmarkItems and
+    // BookmarkFolders have.
+    var bookmarkItemOrFolderTitle: String?
+
+    // Editable field(s) that only BookmarkItems have.
+    var bookmarkItemURL: String?
+
+    var isNew: Bool {
+        return bookmarkNodeGUID == nil
+    }
+
+    var isFolderListExpanded = false
+
+    // Array of tuples containing all of the BookmarkFolders
+    // along with their indentation depth.
+    var bookmarkFolders: [(folder: BookmarkFolder, indent: Int)] = []
+
+    convenience init(profile: Profile, bookmarkNode: BookmarkNode, parentBookmarkFolder: BookmarkFolder) {
+        self.init(profile: profile, bookmarkNodeGUID: bookmarkNode.guid, bookmarkNodeType: bookmarkNode.type, parentBookmarkFolder: parentBookmarkFolder)
+
+        if let bookmarkItem = bookmarkNode as? BookmarkItem {
+            self.bookmarkItemOrFolderTitle = bookmarkItem.title
+            self.bookmarkItemURL = bookmarkItem.url
+
+            self.title = Strings.BookmarksEditBookmark
+        } else if let bookmarkFolder = bookmarkNode as? BookmarkFolder {
+            self.bookmarkItemOrFolderTitle = bookmarkFolder.title
+
+            self.title = Strings.BookmarksEditFolder
+        }
+    }
+
+    convenience init(profile: Profile, withNewBookmarkNodeType bookmarkNodeType: BookmarkNodeType, parentBookmarkFolder: BookmarkFolder) {
+        self.init(profile: profile, bookmarkNodeGUID: nil, bookmarkNodeType: bookmarkNodeType, parentBookmarkFolder: parentBookmarkFolder)
+
+        if bookmarkNodeType == .bookmark {
+            self.bookmarkItemOrFolderTitle = ""
+            self.bookmarkItemURL = ""
+
+            self.title = Strings.BookmarksNewBookmark
+        } else if bookmarkNodeType == .folder {
+            self.bookmarkItemOrFolderTitle = ""
+
+            self.title = Strings.BookmarksNewFolder
+        }
+    }
+
+    private init(profile: Profile, bookmarkNodeGUID: GUID?, bookmarkNodeType: BookmarkNodeType, parentBookmarkFolder: BookmarkFolder) {
+        self.bookmarkNodeGUID = bookmarkNodeGUID
+        self.bookmarkNodeType = bookmarkNodeType
+        self.parentBookmarkFolder = parentBookmarkFolder
+
+        super.init(profile: profile)
+
+        self.tableView.accessibilityIdentifier = "Bookmark Detail"
+        self.tableView.keyboardDismissMode = .onDrag
+        self.tableView.register(TextFieldTableViewCell.self, forCellReuseIdentifier: BookmarkDetailFieldCellIdentifier)
+        self.tableView.register(OneLineTableViewCell.self, forCellReuseIdentifier: BookmarkDetailFolderCellIdentifier)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, closure: { _ in
```

Comment:
nit: trailing closure syntax would be ok here

Reply:
Done.